File: extra_3/app/src/main/java/com/example/extra_3/TaskDao.java
package com.example.extra_3;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface TaskDao {
    @Insert
    void insert(TaskEntity task);

    @Update
    void update(TaskEntity task);

    @Delete
    void delete(TaskEntity task);

    @Query("SELECT * FROM tasks ORDER BY createdAt DESC")
    LiveData<List<TaskEntity>> allByDateDesc();

    @Query("SELECT * FROM tasks ORDER BY title ASC")
    LiveData<List<TaskEntity>> allByTitleAsc();

    @Query("SELECT * FROM tasks WHERE title LIKE '%' || :query || '%' ORDER BY createdAt DESC")
    LiveData<List<TaskEntity>> filterByQuery(String query);
}

File: extra3/app/src/main/java/com/example/extra3/db/TaskDao.kt
package com.example.extra3.db

import androidx.room.*
import kotlinx.coroutines.flow.Flow

@Dao
interface TaskDao {
    @Insert suspend fun insert(task: TaskEntity)
    @Update suspend fun update(task: TaskEntity)
    @Delete suspend fun delete(task: TaskEntity)
    @Query("SELECT * FROM tasks ORDER BY createdAt DESC") fun getAllDesc(): Flow<List<TaskEntity>>
    @Query("SELECT * FROM tasks ORDER BY title ASC") fun getAllByTitle(): Flow<List<TaskEntity>>
}
